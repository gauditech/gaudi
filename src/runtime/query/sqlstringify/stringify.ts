import _ from "lodash";

import { NamePath } from "../build";

import { QueryPlan, QueryPlanExpression, QueryPlanJoin } from "./queryPlan";

import { assertUnreachable } from "@src/common/utils";

function namepathToQuotedPair(npath: NamePath): string {
  return `"${_.initial(npath).join(".")}"."${_.last(npath)}"`;
}

function namepathToQuoted(npath: NamePath): string {
  return `"${npath.join(".")}"`;
}

function exprToString(expr: QueryPlanExpression): string {
  function stringifyOp(lhs: QueryPlanExpression, rhs: QueryPlanExpression, op: string): string {
    return `${exprToString(lhs)} ${op.toUpperCase()} ${exprToString(rhs)}`;
  }
  function stringifyFn(name: string, args: QueryPlanExpression[]): string {
    return `${name}(${args.map(_.unary(exprToString)).join(", ")})`;
  }
  switch (expr.kind) {
    case "alias": {
      return namepathToQuotedPair(expr.value);
    }
    case "literal": {
      return expr.value as string; // FIXME
    }
    case "variable": {
      return `:${expr.name}`;
    }
    case "function": {
      switch (expr.fnName) {
        case "<":
        case ">":
        case ">=":
        case "<=":
        case "/":
        case "*":
        case "in":
        case "not in":
        case "and": {
          return stringifyOp(expr.args[0], expr.args[1], expr.fnName);
        }
        case "+":
        case "-":
        case "or": {
          return `(${stringifyOp(expr.args[0], expr.args[1], expr.fnName)})`;
        }
        case "is": {
          return stringifyOp(expr.args[0], expr.args[1], "=");
        }
        case "is not": {
          return stringifyOp(expr.args[0], expr.args[1], "<>");
        }
        case "length": {
          return stringifyFn("char_length", expr.args);
        }
        case "count":
        case "sum":
        case "lower":
        case "upper":
        case "now":
        case "concat": {
          return stringifyFn(expr.fnName, expr.args);
        }
        case "stringify":
        case "cryptoCompare":
        case "cryptoHash":
        case "cryptoToken":
          throw new Error(`Expression "${expr.fnName}" cannot be used in queries.`);
        default:
          return assertUnreachable(expr.fnName);
      }
    }
  }
}

function joinsToString(joins: QueryPlanJoin[]): string {
  return joins
    .map((join) => {
      switch (join.kind) {
        case "subquery": {
          const subquery = queryPlanToString(join.plan);
          return `
    JOIN (${subquery})
    AS ${namepathToQuoted(join.namePath)}
    ON ${namepathToQuotedPair(join.joinOn[0])} = ${namepathToQuotedPair(join.joinOn[1])}
    ${joinsToString(join.plan.joins)}`;
        }
        case "inline": {
          return `
    JOIN ${join.modelName.toLowerCase()} AS ${namepathToQuoted(join.namePath)}
    ON ${namepathToQuotedPair(join.joinOn[0])} = ${namepathToQuotedPair(join.joinOn[1])}
    ${joinsToString(join.joins)}
    `;
        }
      }
    })
    .join("\n");
}

export function queryPlanToString(plan: QueryPlan): string {
  const selectFrag = plan.select
    ? Object.entries(plan.select)
        .map(([alias, expr]) => `${exprToString(expr)} AS "${alias}"`)
        .join(", ")
    : "*";
  const limitFrag = plan.limit ? `LIMIT ${plan.limit}` : "";
  const orderFrag = plan.orderBy
    ? `ORDER BY ${plan.orderBy.map(
        ([path, dir]): string => `${namepathToQuotedPair(path)} ${dir}`
      )}`
    : "";
  const offsetFrag = plan.offset ? `OFFSET ${plan.offset}` : "";

  const joinFrags = joinsToString(plan.joins);
  const whereFrag = plan.filter ? `WHERE ${exprToString(plan.filter)}` : "";

  const groupByItems = plan.groupBy.map(_.unary(namepathToQuotedPair));
  const groupByFrag = groupByItems.length > 0 ? `GROUP BY ${groupByItems}` : "";

  return `
  SELECT ${selectFrag}
  FROM ${plan.entry.toLowerCase()} AS "${plan.entry}"
  ${joinFrags}
  ${whereFrag}
  ${groupByFrag}
  ${orderFrag}
  ${limitFrag}
  ${offsetFrag} 
  `;
}
