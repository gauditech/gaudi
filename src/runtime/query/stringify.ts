import { source } from "common-tags";
import _ from "lodash";

import { NamePath, selectToSelectable } from "./build";

import { getRef, getTargetModel } from "@src/common/refs";
import { BinaryOperator } from "@src/types/ast";
import {
  Definition,
  FilterDef,
  ModelDef,
  QueryDef,
  QueryDefPath,
  SelectFieldItem,
  SelectableItem,
} from "@src/types/definition";

// FIXME this should accept Queryable
export function queryToString(def: Definition, q: QueryDef): string {
  switch (q.from.kind) {
    case "model": {
      const { value: model } = getRef<"model">(def, q.from.refKey);
      return source`
      SELECT ${selectableToString(def, selectToSelectable(q.select))}
      FROM "${model.dbname}" AS ${namePathToAlias([model.name])}
      ${q.joinPaths.map((j) => joinToString(def, j))}
      WHERE ${filterToString(q.filter)}`;
    }
    case "query":
      return source`
      SELECT ${selectableToString(def, selectToSelectable(q.select))}
      FROM (${queryToString(def, q.from.query)}) AS ${namePathToAlias(q.from.query.fromPath)}
      ${q.joinPaths.map((j) => joinToString(def, j))}
      WHERE ${filterToString(q.filter)}`;
  }
}

function selectableToString(def: Definition, select: SelectableItem[]) {
  return select
    .map((item) => {
      const { value: field } = getRef<"field">(def, item.refKey);
      return `${namePathToAlias(_.initial(item.namePath))}.${field.dbname} AS ${item.alias}`;
    })
    .join(", ");
}

function namePathToAlias(namePath: NamePath): string {
  return `"${namePath.join(".")}"`;
}

function filterToString(filter: FilterDef): string {
  if (filter === undefined) return "TRUE = TRUE";
  switch (filter.kind) {
    case "literal": {
      switch (filter.type) {
        case "boolean":
          return filter.value ? "TRUE" : "FALSE";
        case "null":
          return "NULL";
        case "text":
          return `'${filter.value}'`;
        case "integer":
          return filter.value.toString();
      }
    }
    // Due to bug in eslint/prettier, linter complains that `break` is expected in the case "literal"
    // Since inner switch is exaustive, break is unreachable so prettier deletes it
    // eslint-disable-next-line no-fallthrough
    case "alias": {
      const np = filter.namePath.slice(0, filter.namePath.length - 1);
      const f = filter.namePath.at(filter.namePath.length - 1);
      return `${namePathToAlias(np)}.${f}`;
    }
    case "binary": {
      const fstr = `${filterToString(filter.lhs)} ${opToString(filter.operator)} ${filterToString(
        filter.rhs
      )}`;
      if (filter.operator === "or") {
        return `(${fstr})`;
      }
      return fstr;
    }
    case "variable": {
      return `:${filter.name}`;
    }
  }
}

function opToString(op: BinaryOperator): string {
  switch (op) {
    case "is":
      return "=";
    case "is not":
      return "<>";
    default:
      return op.toUpperCase();
  }
}

function getQuerySource(def: Definition, q: QueryDef): ModelDef {
  switch (q.from.kind) {
    case "model":
      return getRef<"model">(def, q.from.refKey).value;
    case "query": {
      return getQuerySource(def, q.from.query);
    }
  }
}

function joinToString(def: Definition, join: QueryDefPath): string {
  const model = getTargetModel(def.models, join.refKey);
  const joinNames = getJoinNames(def, join.refKey);
  const joinMode = join.joinType === "inner" ? "JOIN" : "LEFT JOIN";
  const joinFilter: FilterDef = {
    kind: "binary",
    operator: "is",
    lhs: { kind: "alias", namePath: [..._.initial(join.namePath), joinNames.from] },
    rhs: { kind: "alias", namePath: [...join.namePath, joinNames.to] },
  };

  let src: string;
  if (join.kind === "query") {
    const { value: query } = getRef<"query">(def, join.refKey);
    const sourceModel = getQuerySource(def, query);
    // extend select
    const conn = mkJoinConnection(sourceModel);

    const retModel = getRef<"model">(def, query.retType).value;
    const fields = retModel.fields.map(
      (f): SelectableItem => ({
        kind: "field",
        refKey: f.refKey,
        alias: f.name,
        name: f.name,
        namePath: [...query.fromPath, f.name],
      })
    );
    src = source`(
      ${queryToString(def, { ...query, select: [...query.select, ...fields, conn] })})`; // FIXME should remove query.select?
  } else {
    src = `"${model.dbname}"`;
  }

  return source`
  ${joinMode}
  ${src} AS ${namePathToAlias(join.namePath)}
  ON ${filterToString(joinFilter)}
  ${join.joinPaths.map((j) => joinToString(def, j))}`;
}

export function mkJoinConnection(model: ModelDef): SelectFieldItem {
  return {
    kind: "field",
    refKey: `${model.name}.id`,
    alias: '"__join_connection"',
    name: "id",
    namePath: [model.name, "id"],
  };
}

function getJoinNames(def: Definition, refKey: string): { from: string; to: string } {
  const prop = getRef(def, refKey);
  switch (prop.kind) {
    case "reference": {
      const reference = prop.value;
      const { value: field } = getRef<"field">(def, reference.fieldRefKey);
      const { value: refField } = getRef<"field">(def, reference.toModelFieldRefKey);

      return { from: field.name, to: refField.name };
    }
    case "relation": {
      const relation = prop.value;
      const { value: reference } = getRef<"reference">(def, relation.throughRefKey);
      const { value: field } = getRef<"field">(def, reference.fieldRefKey);
      const { value: refField } = getRef<"field">(def, reference.toModelFieldRefKey);

      return { from: refField.name, to: field.name };
    }
    case "query": {
      return { from: "id", to: '"__join_connection"' };
    }
    default:
      throw new Error(`Kind ${prop.kind} not supported`);
  }
}
