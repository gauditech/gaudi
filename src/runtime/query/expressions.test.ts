import { queryToString } from "./stringify";

import { compile, compose, parse } from "@src/index";

describe("Expressions to queries", () => {
  it("composes a complex filter expression", () => {
    const bp = `
    model Source {
      relation items { from Item, through source }
      query calc {
        from items
        filter {
          multiplier * (value + 1) / length(concat(textual, "tail")) > 100
        }
      }
    }
    model Item {
      reference source { to Source }
      field value { type integer }
      field multiplier { type integer }
      field textual { type text }
    }
    `;

    const def = compose(compile(parse(bp)));
    const source = def.models.find((m) => m.name === "Source");
    const calc = source!.queries[0]!;
    expect(calc).toMatchSnapshot();
    expect(queryToString(def, calc)).toMatchSnapshot();
  });
  it("composes a complex local computed expression", () => {
    const bp = `
    model Source {
      relation items { from Item, through source }
    }
    model Item {
      reference source { to Source }
      field value { type integer }
      field multiplier { type integer }
      field textual { type text }
      computed worthiness {
        multiplier * (value + 1) / length(concat(textual, "tail"))
      }
    }
    `;

    const def = compose(compile(parse(bp)));
    const item = def.models.find((m) => m.name === "Item");
    const worthiness = item!.computeds[0]!;
    expect(worthiness).toMatchSnapshot();
  });
});
