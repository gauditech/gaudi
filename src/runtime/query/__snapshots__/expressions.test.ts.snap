// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aggregates to queries composes a query with simple aggregate through query 1`] = `
"SELECT
  "Org"."id" AS "id",
  "Org"."name" AS "name",
  "Org.repo_count"."result" AS "repo_count"
FROM
  "org" AS "Org"
  LEFT JOIN (
    SELECT
      "Org".id,
      count("Org.all_repos_nested".id) AS "result"
    FROM
      "org" AS "Org"
      JOIN (
        SELECT
          "Org.all_repos"."id" AS "id",
          "Org.all_repos"."name" AS "name",
          "Org.all_repos"."org_id" AS "org_id",
          "Org"."id" AS "__join_connection"
        FROM
          "org" AS "Org"
          JOIN (
            SELECT
              "Org.repos"."id" AS "id",
              "Org.repos"."name" AS "name",
              "Org.repos"."org_id" AS "org_id",
              "Org"."id" AS "__join_connection"
            FROM
              "org" AS "Org"
              JOIN "repo" AS "Org.repos" ON "Org"."id" = "Org.repos"."org_id"
          ) AS "Org.all_repos" ON "Org"."id" = "Org.all_repos"."__join_connection"
      ) AS "Org.all_repos_nested" ON "Org"."id" = "Org.all_repos_nested"."__join_connection"
    GROUP BY
      "Org".id
  ) AS "Org.repo_count" ON "Org".id = "Org.repo_count".id"
`;

exports[`Aggregates to queries composes a query with simple aggregate through relation 1`] = `
"SELECT
  "Org"."id" AS "id",
  "Org"."name" AS "name",
  "Org.repo_count"."result" AS "repo_count"
FROM
  "org" AS "Org"
  LEFT JOIN (
    SELECT
      "Org".id,
      count("Org.repos".id) AS "result"
    FROM
      "org" AS "Org"
      JOIN "repo" AS "Org.repos" ON "Org"."id" = "Org.repos"."org_id"
    GROUP BY
      "Org".id
  ) AS "Org.repo_count" ON "Org".id = "Org.repo_count".id"
`;

exports[`Expressions to queries composes a complex filter expression 1`] = `
{
  "filter": {
    "args": [
      {
        "args": [
          {
            "args": [
              {
                "kind": "alias",
                "namePath": [
                  "Source",
                  "items",
                  "multiplier",
                ],
              },
              {
                "args": [
                  {
                    "kind": "alias",
                    "namePath": [
                      "Source",
                      "items",
                      "value",
                    ],
                  },
                  {
                    "kind": "literal",
                    "type": "integer",
                    "value": 1,
                  },
                ],
                "kind": "function",
                "name": "+",
              },
            ],
            "kind": "function",
            "name": "*",
          },
          {
            "args": [
              {
                "args": [
                  {
                    "kind": "alias",
                    "namePath": [
                      "Source",
                      "items",
                      "textual",
                    ],
                  },
                  {
                    "kind": "literal",
                    "type": "text",
                    "value": "tail",
                  },
                ],
                "kind": "function",
                "name": "concat",
              },
            ],
            "kind": "function",
            "name": "length",
          },
        ],
        "kind": "function",
        "name": "/",
      },
      {
        "kind": "literal",
        "type": "integer",
        "value": 100,
      },
    ],
    "kind": "function",
    "name": ">",
  },
  "fromPath": [
    "Source",
    "items",
  ],
  "kind": "query",
  "modelRefKey": "Source",
  "name": "calc",
  "refKey": "Source.calc",
  "retType": "Item",
  "select": [],
}
`;

exports[`Expressions to queries composes a complex filter expression 2`] = `
"SELECT
FROM
  "source" AS "Source"
  JOIN "item" AS "Source.items" ON "Source"."id" = "Source.items"."source_id"
WHERE
  "Source.items"."multiplier" * ("Source.items"."value" + 1) / char_length(concat("Source.items"."textual", 'tail')) > 100"
`;

exports[`Expressions to queries composes a complex local computed expression 1`] = `
{
  "exp": {
    "args": [
      {
        "args": [
          {
            "kind": "alias",
            "namePath": [
              "Item",
              "multiplier",
            ],
          },
          {
            "args": [
              {
                "kind": "alias",
                "namePath": [
                  "Item",
                  "value",
                ],
              },
              {
                "kind": "literal",
                "type": "integer",
                "value": 1,
              },
            ],
            "kind": "function",
            "name": "+",
          },
        ],
        "kind": "function",
        "name": "*",
      },
      {
        "args": [
          {
            "args": [
              {
                "kind": "alias",
                "namePath": [
                  "Item",
                  "textual",
                ],
              },
              {
                "kind": "literal",
                "type": "text",
                "value": "tail",
              },
            ],
            "kind": "function",
            "name": "concat",
          },
        ],
        "kind": "function",
        "name": "length",
      },
    ],
    "kind": "function",
    "name": "/",
  },
  "kind": "computed",
  "modelRefKey": "Item",
  "name": "worthiness",
  "refKey": "Item.worthiness",
}
`;

exports[`Expressions to queries composes a correct wrapped query which contains local computed prop 1`] = `
{
  "filter": {
    "args": [
      {
        "kind": "alias",
        "namePath": [
          "Source",
          "items",
          "worthiness",
        ],
      },
      {
        "kind": "literal",
        "type": "integer",
        "value": 100,
      },
    ],
    "kind": "function",
    "name": ">",
  },
  "fromPath": [
    "Source",
    "items",
  ],
  "kind": "query",
  "modelRefKey": "Source",
  "name": "calc",
  "refKey": "Source.calc",
  "retType": "Item",
  "select": [
    {
      "alias": "id",
      "kind": "field",
      "name": "id",
      "namePath": [
        "Source",
        "items",
        "id",
      ],
      "refKey": "Item.id",
    },
    {
      "alias": "value",
      "kind": "field",
      "name": "value",
      "namePath": [
        "Source",
        "items",
        "value",
      ],
      "refKey": "Item.value",
    },
    {
      "alias": "multiplier",
      "kind": "field",
      "name": "multiplier",
      "namePath": [
        "Source",
        "items",
        "multiplier",
      ],
      "refKey": "Item.multiplier",
    },
    {
      "alias": "textual",
      "kind": "field",
      "name": "textual",
      "namePath": [
        "Source",
        "items",
        "textual",
      ],
      "refKey": "Item.textual",
    },
    {
      "alias": "worthiness",
      "kind": "computed",
      "name": "worthiness",
      "namePath": [
        "Source",
        "items",
        "worthiness",
      ],
      "refKey": "Item.worthiness",
    },
    {
      "alias": "text_tail_len",
      "kind": "computed",
      "name": "text_tail_len",
      "namePath": [
        "Source",
        "items",
        "text_tail_len",
      ],
      "refKey": "Item.text_tail_len",
    },
  ],
}
`;

exports[`Expressions to queries composes a correct wrapped query which contains local computed prop 2`] = `
"SELECT
  "Source.items"."id" AS "id",
  "Source.items"."value" AS "value",
  "Source.items"."multiplier" AS "multiplier",
  "Source.items"."textual" AS "textual",
  (
    (
      "Source.items"."multiplier" * ("Source.items"."value" + 1) / char_length(concat("Source.items"."textual", 'tail')) + 10
    ) + "Source.items.source.total_items"."result"
  ) AS "worthiness",
  char_length(concat("Source.items"."textual", 'tail')) AS "text_tail_len"
FROM
  "source" AS "Source"
  JOIN "item" AS "Source.items" ON "Source"."id" = "Source.items"."source_id"
  JOIN "source" AS "Source.items.source" ON "Source.items"."source_id" = "Source.items.source"."id"
  LEFT JOIN (
    SELECT
      "Source".id,
      count("Source.items".id) AS "result"
    FROM
      "source" AS "Source"
      JOIN "item" AS "Source.items" ON "Source"."id" = "Source.items"."source_id"
    GROUP BY
      "Source".id
  ) AS "Source.items.source.total_items" ON "Source.items.source".id = "Source.items.source.total_items".id
WHERE
  (
    (
      "Source.items"."multiplier" * ("Source.items"."value" + 1) / char_length(concat("Source.items"."textual", 'tail')) + 10
    ) + "Source.items.source.total_items"."result"
  ) > 100"
`;
